### ISSUE TITLE
record inference result

### ISSUE THREAD

**Issue by GY0913** [NVIDIA/DALI#1145]
hello，i use HybridPipeline to read images, set 4 pipe ,but when i write done the test result during one epoch loop only have 1/4 results in the log. is there any thing wrong to use like this way?

**Comment by awolant** [comment]
Hi, thanks for the question. Maybe you could provide some minimal reproduction for what you are trying to do?
From the top of my head, maybe you are writing to the same paths from all of the pipelines?

**Comment by GY0913** [comment]
`    model_names = ['5000.pth.tar','1000.pth.tar',
                    '15-0.pth.tar','18-3000.pth.tar']

    pipes  = [HybridTrainPipe(file_list='' batch_size=32, num_threads=16, 
             device_id = device_id, data_dir='', num_gpus = 4) for device_id in range(4)]
    pipes[0].build()
    loader = DALIGenericIterator(pipes, ['data', 'label'], pipes[0].epoch_size("Reader"))

    for model_name_ in model_names:
        model = ..   # create model
        model.eval()
        with torch.no_grad():
            path_list = []
            score = []
            for batch_idx, data_ in enumerate(loader):              
                for d in data_:                   
                    input = d["data"]
                    path = d["label"].squeeze().cuda().long()
                input = input.cuda(args.gpu, non_blocking=True)
                output = model(input)

                probability = F.softmax(output,dim = 1)
                score.extend(probability)
                path_list.extend(path)

            with open(os.path.join(args.output_dir,savename),'w') as ff:
            	'write down path_list and score'

            loader.reset()`

only 1/4 of loader._size() in final result.log

**Comment by awolant** [comment]
@GY0913 In your code example `data_` is a list of length 4 returned from the iterator (one output from each of the pipelines). It seems that you iterate over this list (in `for d in data_:`), but override the value of `input` on every `d`, so by the time you invoke your model you end up only with the data from the last pipeline. The solution in this case would be either to extend the input or to call the model on every iteration of the most inner loop, I think.

**Comment by GY0913** [comment]
@awolant so when i set 4 pipes in training , actually 1/4 data is learned in one epoch ? how can i fix this   in training ? can you provide some examples like resnet50-example in doc (which only use one pipe)

**Comment by GY0913** [comment]
`for batch_idx, data_ in enumerate(loader):
                print(len(data_),data_)`
I print data_ and find out that it has the same output from each of the pipelines ?
 confuse about how multi-pipes work

**Comment by awolant** [comment]
@GY0913 As far as I can tell from this code example, when you create your pipelines, the only difference between them is the `device_id`. That means they will run on different GPUs, but on the same data. That might be why you see the same data when printing it.
In DALI, to control this behavior you can use `shard_id` and `num_shards` parameters of the [readers](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/supported_ops.html?highlight=filereader#nvidia.dali.ops.FileReader). In this particular case, you have 4 shards and each pipeline should work on a different shard, so you need to pass different `shard_id` (much like you do with `device_id`). Are you sure you are using those parameters correctly?
Regarding the training with multiple devices, you can refer to our [PyTorch ResNet example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/pytorch/resnet50/pytorch-resnet50.html) or [MXNet ResNet example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/mxnet-resnet50.html) depending on the framework you use. They show how to setup full training on multiple devices with code examples.
If you have more questions, please ask.

**Comment by GY0913** [comment]
I set  [shard_id=device_id, num_shards=num_gpus(which is 4 in my case)]  in FileReader, as far as I understand the dataset is splited into num_gpus parts ,and each pipe read 1/4 part of the whole dataset(each pipe read different data),but when I print out data_ ,it is the same data repeat 4 time ,why pipes not reading differently? not sure what's wrong

**Comment by awolant** [comment]
@GY0913 can you provide more complete reproduction, so I can take a look?

**Comment by GY0913** [comment]
thanks for the replies. somehow I finger it out . guess my code is not updating to server successfully during the last few hours ,running the old version makes a lot mistakes.
 thanks again for your help!

**Comment by awolant** [comment]
@GY0913 Great, happy to help. Closing for now.

**Comment by GY0913** [comment]
`batch_num = int((train_dataloader._size )/(conf.batch_sizes*4))  
for epoch in range(conf.start_epoch, conf.epoch):
    model.train()
    scheduler.step()
    accuracy.reset()

    running_loss = 0.0
    print(datetime.now())
    for i, data in enumerate(train_dataloader):
        for d in data:
            classes = d["label"].squeeze().cuda().long()
            inputs = d["data"]
        inputs = inputs.to(device)
        classes = classes.to(device)

        optimizer.zero_grad()
        inputs = inputs.float()

        cls_out = model(inputs)
        loss = criterion(cls_out, classes)

        loss.backward()
        optimizer.step()

        # print statistics
        running_loss += loss.item()
        accuracy.add(cls_out.detach().cpu().numpy(), classes)`
still a little confuse about the training TAT
I use caffe-lmdb data in training and  modify the code from : https://github.com/NVIDIA/DALI/blob/b9d05de59608ef404bf8733a8de297e3fcd8054e/docs/examples/pytorch/pytorch-various-readers.ipynb
so my code of loading data to train is  correct  ?

**Comment by awolant** [comment]
@GY0913 In the code above you still override the `inputs` variable.
If you are using PyTorch, maybe you could look at our [ResNet training example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/pytorch/resnet50/pytorch-resnet50.html).
It shows how to setup full training on multigpu.

**Comment by GY0913** [comment]
thanks for the advice !
